Add handling for synchronized low power tickers

c1728p9 opened the pull request:
Some low power tickers take multiple cycles of the low power clock to set a compare value. Because of this if the compare value is set twice back-to-back these implementations will block until that time has passed. This can cause system stability issues since interrupts are disabling for this time.

To gracefully support this kind of hardware this patch adds code to prevent back-to-back writes to the hardware. It does this by recording the low power clock cycle of the initial write. If any writes come in too soon after this initial write the microsecond ticker is used to schedule the new write in the future when the hardware is ready to accept a new value.

To enable this feature on a target the macro `LOWPOWERTIMER_DELAY_TICKS` must be set to the number of low power clock cycles that must elapse between writes to the low power timer.

c1728p9 commented:
CC @jeromecoutant @LMESTM

0xc0170 commented:
Shouldn't `LOWPOWERTIMER_DELAY_TICKS ` be part of ticker info? It would be better to have that member there than macro?
Minar had some similar requirement for tickers - the nearest delay that a ticker can schedule - if value is lower than that, it would be fired immediately. I think this is similar concept here, isn't it ?

c1728p9 commented:
Updated this PR by making the following changes:
- Moved wrapper code to a new file
- Changed mbed_lp_ticker_wrapper back to .c (from .cpp)
- Removed macro for set_interrupt and instead using #if defined directly.

c1728p9 commented:
> Shouldn't LOWPOWERTIMER_DELAY_TICKS be part of ticker info? It would be better to have that member there than macro?

I made this enabled by a macro so this code won't be present for devices which don't need it. As an alternative it could be a part of ticker info but in that case when you use the low power ticker, the microsecond ticker will also be included in your binary even for devices which don't need it.

> if value is lower than that, it would be fired immediately

There is already the fire now ticker API which can be used to trigger interrupts immediately, so I'm not sure if there would be much value to this behavior. **With the current implementation the code guarantees that the ticker irq will fire at or after the tick specified.** Firing early would break this guarantee and make it harder to ensure correct handling in the common ticker layer.

> Minar had some similar requirement for tickers - the nearest delay that a ticker can schedule - if value is lower than that, it would be fired immediately. I think this is similar concept here, isn't it ?

The primary purpose of this change is to prevent blocking when set_interrupt is called back-to-back, regardless of the value being set. Calling set interrupt to fire 1 second in the future and then calling it again immediately to fire 2 seconds in the future causes blocking without this change. Minar like requirements is also added as a side effect -

In addition to the code to prevent blocking, this patch also ensure that lp_ticker_set_interrupt isn't called with an alarm time sooner than LOWPOWERTIMER_DELAY_TICKS. This is done because the value LOWPOWERTIMER_DELAY_TICKS implies that it takes that amount of time for the value to take effect - If you set it to a value less than this it may cause the event to trigger on the next rollover rather than in a few clock cycles.

If these values aren't intrinsically coupled (which they may be), it may be best to separate out these two concepts so they aren't conflated:
1. ```LOWPOWERTIMER_DELAY_TICKS``` - Number of clock cycles that must pass between calls to ```ticker_set_interrupt```
1. The minimum number of clock cycles in the future that it is safe to setup an alarm with ```ticker_set_interrupt```

What do you think @0xc0170, @kjbracey-arm, @LMESTM, @jeromecoutant, @pan-?

c1728p9 commented:
/morph build

0xc0170 commented:
> What do you think @0xc0170, @kjbracey-arm, @LMESTM, @jeromecoutant, @pan-?

LGTM, so delay low power ticks stays as it is in this PR

0xc0170 commented:
/morph test

cmonr commented on hal/mbed_lp_ticker_wrapper.cpp:
```diff
@@ -0,0 +1,142 @@
+ */
+#include "hal/lp_ticker_api.h"
+
+#if DEVICE_LOWPOWERTIMER && defined(LOWPOWERTIMER_DELAY_TICKS) && (LOWPOWERTIMER_DELAY_TICKS > 0)
```
Is there a reason why `DEVICE_LOWPOWERTIMER` isn't wrapped with a `defined`?

0xc0170 replied:
would not hurt but by default values not set defaults to 0.

cmonr commented on hal/mbed_lp_ticker_wrapper.cpp:
```diff
@@ -0,0 +1,142 @@
+        next = timestamp;
+        last_request = current;
+        if (!timeout_pending) {
+            timeout->attach_us(set_interrupt_later, reschedule_us);
```
I take it that this doesn't need to be unattached once the timeout elapses?

0xc0170 replied:
Not needed. Detach only if you want to remove it early

0xc0170 commented:
I would like to see the feedback from HAL, please review

c1728p9 commented:
I removed the unnecessary define() checks.

c1728p9 commented:
/morph build

pan- commented on hal/mbed_lp_ticker_wrapper.cpp:
```diff
@@ -0,0 +1,142 @@
+static timestamp_t reschedule_us;
+
+// Do not use SingletonPtr since this must be initialized in a critical section
+static mbed::Timeout *timeout;
```
Is it correct to use a C++ driver and C++ code in the HAL ?

c1728p9 replied:
There is no mbed-os rule or technical reason preventing C++ from being used in the HAL. Most of the HAL drivers just don't need to be. In this case there is no API with this functionality in the C layer, so I used the C++ Timeout class.

pan- commented on hal/mbed_lp_ticker_wrapper.cpp:
```diff
@@ -0,0 +1,142 @@
+static timestamp_t next = 0;
+
+static timestamp_t mask;
+static timestamp_t min_delta;
```
This can be a constant assigned with the value LOWPOWERTIMER_DELAY_TICKS

c1728p9 replied:
done

pan- commented on hal/lp_ticker_api.h:
```diff
@@ -48,6 +48,23 @@ typedef void (*ticker_irq_handler_type)(const ticker_data_t *const);
+ * @param timestamp Time to call ticker irq
+ * @note this is a utility function and it's not required part of HAL implementation
+ */
+void lp_ticker_set_interrupt_wrapper(timestamp_t timestamp);
```
There's no reason to expose this function in the HAL interface file; this can remain hidden. 
However it would be valuable to improve documentation of `lp_ticker_set_interrupt` by indicating that if `LOWPOWERTIMER_DELAY_TICKS` is greater than 0 then this mechanism exist and the wheel should not be reinvented.

c1728p9 replied:
I removed this from the header to reduce its scope.

Also I created a docs PR here:
https://github.com/ARMmbed/Handbook/pull/491

cmonr commented:
/morph test

0xc0170 commented:
I did not get full details for the issue in https://github.com/ARMmbed/mbed-os/pull/6522.  But to me it seems related (please review). Realtek has a ticker (in this case high freq) also that takes some time to set interrupt.

cc @M-ichae-l

c1728p9 commented:
I pulled the fix https://github.com/ARMmbed/mbed-os/pull/6473/commits/436f1d1e76096faf8d1791abac9469033e0eff5c into this PR.

cmonr commented:
/morph build

0xc0170 commented:
Waiting for the @kjbracey-arm @bulislaw review

cmonr commented:
@bulislaw Any comments?
